testarch/utils/testarch_graph.py

```diff
@@ -1,140 +1,174 @@
+                            continue
+                        tracks_for_pin = ytracks[loc.x]
+                    elif pin_side == tracks.Direction.TOP:
+                        assert pin.name == 'BLK_TI-TILE.COUT[0]', pin
+                        continue
+                    elif pin_side == tracks.Direction.BOTTOM:
+                        assert pin.name == 'BLK_TI-TILE.CIN[0]', pin
+                        continue
+                    else:
+                        assert False, pin_side
+
+                    if pin_class.type == graph2.PinType.OUTPUT:
+                        for track_inode in tracks_for_pin:
+                            graph.add_edge(
+                                    src_node=pin_node,
+                                    sink_node=track_inode,
+                                    switch_id=switch,
+                                    )
+                    elif pin_class.type == graph2.PinType.INPUT:
+                        for track_inode in tracks_for_pin:
+                            graph.add_edge(
+                                    src_node=track_inode,
+                                    sink_node=pin_node,
+                                    switch_id=switch,
+                                    )
+
+
+def connect_tracks_to_tracks(graph, switch, verbose=False):
     print("Connecting tracks to tracks")
 
-    # Iterate over all valid x channels and connect to all valid y channels and vice versa as direction implies
-    for xtrack in g.channels.x.tracks():
-        for ytrack in g.channels.y.tracks():
-
-            def try_connect(atrack, btrack):
-                # One of the nodes should be going in and the other should be going out
-                # Filter out grossly non-sensical connections that confuse the VPR GUI
-                if (btrack.direction == channel.Track.Direction.INC
-                        and btrack.start0 <= atrack.common
-                        or btrack.direction == channel.Track.Direction.DEC
-                        and btrack.start0 > atrack.common):
-                    g.connect_track_to_track(btrack, atrack, switch)
-                    if verbose:
-                        print("Connect %s to %s" % (btrack, atrack))
-
-            try_connect(xtrack, ytrack)
-            try_connect(ytrack, xtrack)
-
+    # Iterate over all valid x channels and connect to all valid y channels
+    # and vice versa as direction implies
+
+    def try_connect(ainode, binode):
+        atrack = graph.nodes[ainode]
+        btrack = graph.nodes[binode]
+        # One of the nodes should be going in and the other should be going out
+        # Filter out grossly non-sensical connections that confuse the VPR GUI
+        if (btrack.direction == graph2.NodeDirection.INC_DIR
+                and channel_start(btrack) <= channel_common(atrack)) \
+            or (btrack.direction == graph2.NodeDirection.DEC_DIR
+                and channel_start(btrack) > channel_common(atrack)):
```

Comment:
I know it's just a translation of the previous implementation
```suggestion
                and channel_start(btrack) >= channel_common(atrack)):
```
Not sure why you connect `DEC_DIR` only from either x->y or y->x exclusively for non-square grids.

Reply:
Totally valid, fixed.